src/mlpack/methods/ann/loss_functions/reconstruction_loss.hpp

```diff
@@ -0,0 +1,89 @@
+/**
+ * @file reconstruction_loss.hpp
+ * @author Atharva Khandait
+ *
+ * Definition of the reconstruction loss performance function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_RECONSTRUCTION_LOSS_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_RECONSTRUCTION_LOSS_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/dists/normal_distribution.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The mean squared error performance function measures the network's
+ * performance according to the mean of squared errors.
+ *
+ * @tparam ActivationFunction Activation function used for the embedding layer.
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat,
+    typename DistType = NormalDistribution<InputDataType>
+>
+class ReconstructionLoss
+{
+ public:
+  /**
+   * Create the ReconstructionLoss object.
+   */
+  ReconstructionLoss();
+
+  /*
+   * Computes the mean squared error function.
+   *
+   * @param input Input data used for evaluating the specified function.
+   * @param output Resulting output activation.
+   */
+  template<typename InputType, typename TargetType>
+  double Forward(const InputType&& input, const TargetType&& target);
+  /**
+   * Ordinary feed backward pass of a neural network.
+   *
+   * @param input The propagated input activation.
+   * @param target The target vector.
+   * @param output The calculated error.
+   */
+  template<typename InputType, typename TargetType, typename OutputType>
+  void Backward(const InputType&& input,
+                const TargetType&& target,
+                OutputType&& output);
+
+  //! Get the output parameter.
+  OutputDataType& OutputParameter() const { return outputParameter; }
+  //! Modify the output parameter.
+  OutputDataType& OutputParameter() { return outputParameter; }
+
+  /**
+   * Serialize the layer
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored distribution object.
+  DistType* dist;
```

Comment:
Sorry if this doubt seems silly, but how would I use the constructor in reconstruction_loss_impl.hpp if I already declare 
```
DistType dist;
```
 It will just use the empty default constructor which I don't want to use.

Reply:
Maybe this was forgotten, can you please take a look at this?

Reply:
Sorry, I hadn't had a chance to get around to it yet (but I should have, sorry about that).  You can just use the constructor initialization list to do this:

```
ReconstructionLoss::ReconstructionLoss() :
    dist(/* arguments to constructor */)
{
  ...
}
```

For the other code where you run `dist = new DistType(std::move(input));` you can just do `dist = DistType(std::move(input));`.

Reply:
Oh sorry, that was really silly. I don’t know why I got confused over that.

Reply:
No worries, happy to help :)